sstables/sstable_directory.cc

```diff
@@ -244,16 +244,17 @@ sstable_directory::process_sstable_dir(process_flags flags) {
             throw sstables::malformed_sstable_exception(format("At directory: {}: no TOC found for SSTable {}!. Refusing to boot", _sstable_dir.native(), path.native()));
         } else {
             dirlog.info("Found incomplete SSTable {} at directory {}. Removing", path.native(), _sstable_dir.native());
-            _files_for_removal.insert(path.native());
+            state.files_for_removal.insert(path.native());
         }
     }
 }
 
 future<>
 sstable_directory::commit_directory_changes() {
+    auto files_for_removal = std::move(_lister->_state->files_for_removal);
     _lister.reset();
     // Remove all files scheduled for removal
-    return parallel_for_each(std::exchange(_files_for_removal, {}), [] (sstring path) {
+    return parallel_for_each(std::move(files_for_removal), [] (sstring path) {
```

Comment:
parallel_for_each creates all futures before yielding so no need to `std::move(files_for_removal)`.
I'd consider to convert this function to a coroutine to make it even clearer that `files_for_removal` remains in the coroutine scope until the parallel loop future is resolved:
```c++
    co_await coroutine::parallel_for_each(files_for_removal, [] (const sstring& path) {
        dirlog.info("Removing file {}", path);
        return remove_file(path);
    });
```

Reply:
Good point. Will queue this followup